File: js/noisemaker/dsl/tokenizer.js
// Support ternary operator and equals sign for named arguments
const PUNCT = new Set(['(', ')', '{', '}', '[', ']', ',', ':', '.', '+', '-', '*', '/', '?', '=', '<', '>']);

function isDigit(ch) {
  return ch >= '0' && ch <= '9';
}

function isHex(ch) {
  return isDigit(ch) || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F');
}

function isIdentStart(ch) {
  return (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') || ch === '_';
}

function isIdent(ch) {
  return isIdentStart(ch) || isDigit(ch);
}

function makeError(message, line, column) {
  const err = new Error(`${message} at line ${line} column ${column}`);
  err.line = line;
  err.column = column;
  return err;
}

/**
 * Convert source string to a list of tokens.
 *
 * Each token has shape: { type, value, line, column }
 *
 * @param {string} source
 * @returns {Array<object>}
 */
export function tokenize(source) {
  const tokens = [];
  let i = 0;
  let line = 1;
  let column = 1;

  const length = source.length;

  function peek(offset = 0) {
    return source[i + offset];
  }

  function advance() {
    const ch = source[i++];
    if (ch === '\n') {
      line++;
      column = 1;
    } else {
      column++;
    }
    return ch;
  }

  function skipWhitespace() {
    while (i < length) {
      const ch = peek();
      if (ch === ' ' || ch === '\t' || ch === '\r' || ch === '\n') {
        advance();
        continue;
      }
      if (ch === '/' && peek(1) === '/') {
        advance();
        advance();
        while (i < length && peek() !== '\n') {
          advance();
        }
        continue;
      }
      if (ch === '/' && peek(1) === '*') {
        const commentLine = line;
        const commentColumn = column;
        advance();
        advance();
        let closed = false;
        while (i < length) {
          if (peek() === '*' && peek(1) === '/') {
            advance();
            advance();
            closed = true;
            break;
          }
          advance();
        }
        if (!closed) {
          throw makeError('Unterminated multi-line comment', commentLine, commentColumn);
        }
        continue;
      }
      break;
    }
  }

  while (i < length) {
    skipWhitespace();
    if (i >= length) break;

    const ch = peek();
    const tokenLine = line;
    const tokenColumn = column;

    // Number literal (support optional leading dot)
    if (isDigit(ch) || (ch === '.' && isDigit(peek(1)))) {
      let numStr = '';
      if (ch === '.') {
        numStr = '0';
        advance();
      }
      while (isDigit(peek())) {
        numStr += advance();
      }
      if (peek() === '.') {
        numStr += advance();
        while (isDigit(peek())) {
          numStr += advance();
        }
      }
      tokens.push({
        type: 'number',
        value: parseFloat(numStr),
        line: tokenLine,
        column: tokenColumn,
      });
      continue;
    }

    // String literal (single or double quotes)
    if (ch === '"' || ch === '\'') {
      const quote = advance();
      let str = '';
      while (i < length && peek() !== quote) {
        const c = advance();
        if (c === '\\') {
          if (i >= length) {
            throw makeError('Unterminated string literal', tokenLine, tokenColumn);
          }
          const esc = advance();
          const map = { n: '\n', r: '\r', t: '\t', '\\': '\\', '"': '"', '\'': '\'' };
          str += map[esc] !== undefined ? map[esc] : esc;
        } else if (c === '\n') {
          throw makeError('Unterminated string literal', tokenLine, tokenColumn);
        } else {
          str += c;
        }
      }
      if (peek() !== quote) {
        throw makeError('Unterminated string literal', tokenLine, tokenColumn);
      }
      advance();
      tokens.push({
        type: 'string',
        value: str,
        line: tokenLine,
        column: tokenColumn,
      });
      continue;
    }

    // Hex colour literal
    if (ch === '#') {
      advance();
      let hex = '';
      while (i < length && isHex(peek())) {
        hex += advance();
      }
      if (hex.length !== 3 && hex.length !== 6) {
        throw makeError('Invalid color', tokenLine, tokenColumn);
      }
      tokens.push({
        type: 'color',
        value: `#${hex}`,
        line: tokenLine,
        column: tokenColumn,
      });
      continue;
    }

    // Punctuation
    if (PUNCT.has(ch)) {
      advance();
      tokens.push({
        type: ch,
        value: ch,
        line: tokenLine,
        column: tokenColumn,
      });
      continue;
    }

    // Identifier / keywords
    if (isIdentStart(ch)) {
      let id = '';
      while (i < length && isIdent(peek())) {
        id += advance();
      }
      if (id === 'true' || id === 'false') {
        tokens.push({
          type: 'boolean',
          value: id === 'true',
          line: tokenLine,
          column: tokenColumn,
        });
      } else if (id === 'null') {
        tokens.push({
          type: 'null',
          value: null,
          line: tokenLine,
          column: tokenColumn,
        });
      } else {
        tokens.push({
          type: 'identifier',
          value: id,
          line: tokenLine,
          column: tokenColumn,
        });
      }
      continue;
    }

    // Unrecognised character
    throw makeError(`Unexpected character '${ch}'`, tokenLine, tokenColumn);
  }

  return tokens;
}

File: noisemaker/dsl/tokenizer.py
# Support ternary operator and equals sign for named arguments
PUNCT = set("(){}[],.:+-*/?=<>")


class TokenizerError(Exception):
    """Custom exception for tokenizer errors with line and column tracking."""

    def __init__(self, message: str, line: int, column: int):
        super().__init__(f"{message} at line {line} column {column}")
        self.line = line
        self.column = column


def is_digit(ch):
    return ch is not None and "0" <= ch <= "9"


def is_hex(ch):
    return ch is not None and (is_digit(ch) or ("a" <= ch <= "f") or ("A" <= ch <= "F"))


def is_ident_start(ch):
    return ("A" <= ch <= "Z") or ("a" <= ch <= "z") or ch == "_"


def is_ident(ch):
    return is_ident_start(ch) or is_digit(ch)


def make_error(message, line, column):
    return TokenizerError(message, line, column)


def tokenize(source):
    tokens = []
    i = 0
    line = 1
    column = 1
    length = len(source)

    def peek(offset=0):
        idx = i + offset
        return source[idx] if idx < length else None

    def advance():
        nonlocal i, line, column
        ch = source[i]
        i += 1
        if ch == "\n":
            line += 1
            column = 1
        else:
            column += 1
        return ch

    def skip_whitespace():
        nonlocal i
        while i < length:
            ch = peek()
            if ch in (" ", "\t", "\r", "\n"):
                advance()
                continue
            if ch == "/" and peek(1) == "/":
                advance()
                advance()
                while i < length and peek() != "\n":
                    advance()
                continue
            if ch == "/" and peek(1) == "*":
                comment_line = line
                comment_column = column
                advance()
                advance()
                closed = False
                while i < length:
                    if peek() == "*" and peek(1) == "/":
                        advance()
                        advance()
                        closed = True
                        break
                    advance()
                if not closed:
                    raise make_error("Unterminated multi-line comment", comment_line, comment_column)
                continue
            break

    while i < length:
        skip_whitespace()
        if i >= length:
            break
        ch = peek()
        token_line = line
        token_column = column
        if is_digit(ch) or (ch == "." and is_digit(peek(1))):
            num_str = ""
            if ch == ".":
                num_str = "0"
                advance()
            while is_digit(peek()):
                num_str += advance()
            if peek() == ".":
                num_str += advance()
                while is_digit(peek()):
                    num_str += advance()
            tokens.append({"type": "number", "value": float(num_str), "line": token_line, "column": token_column})
            continue
        if ch in ('"', "'"):
            quote = advance()
            s = ""
            while i < length and peek() != quote:
                c = advance()
                if c == "\\":
                    if i >= length:
                        raise make_error("Unterminated string literal", token_line, token_column)
                    esc = advance()
                    mapping = {"n": "\n", "r": "\r", "t": "\t", "\\": "\\", '"': '"', "'": "'"}
                    s += mapping.get(esc, esc)
                elif c == "\n":
                    raise make_error("Unterminated string literal", token_line, token_column)
                else:
                    s += c
            if peek() != quote:
                raise make_error("Unterminated string literal", token_line, token_column)
            advance()
            tokens.append({"type": "string", "value": s, "line": token_line, "column": token_column})
            continue
        if ch == "#":
            advance()
            hex_str = ""
            while i < length and is_hex(peek()):
                hex_str += advance()
            if len(hex_str) not in (3, 6):
                raise make_error("Invalid color", token_line, token_column)
            tokens.append({"type": "color", "value": f"#{hex_str}", "line": token_line, "column": token_column})
            continue
        if ch in PUNCT:
            advance()
            tokens.append({"type": ch, "value": ch, "line": token_line, "column": token_column})
            continue
        if is_ident_start(ch):
            ident = ""
            while i < length and is_ident(peek()):
                ident += advance()
            if ident in ("true", "false"):
                tokens.append({"type": "boolean", "value": ident == "true", "line": token_line, "column": token_column})
            elif ident == "null":
                tokens.append({"type": "null", "value": None, "line": token_line, "column": token_column})
            else:
                tokens.append({"type": "identifier", "value": ident, "line": token_line, "column": token_column})
            continue
        raise make_error(f"Unexpected character '{ch}'", token_line, token_column)
    return tokens
